### PR TITLE
Feature/expression api

### PR DESCRIPTION
**Pull request by pnemere**

### CODE REVIEW

**Review by RyanStonebraker — APPROVED** [review]
A few minor things, but I was able to get everything working on my end! I think this will be quite a bit easier to test from the UI with the new code editor merged in too, so going to pull this branch into mine and work on integration. Feel free to merge when ready though!

File: client/src/app/expression-language/lua/Map.lua [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                    mapMax = v[2]
+                end
+            end
+        end
+    end
+    return mapMin, mapMax
+end
+
+
+function Map.normalise(m)
```
[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/Map.lua:347]: Good point, though I don't know if the function was ever used, maybe it's less work to just remove it 😆

File: client/src/app/expression-language/lua/test/Map-test.lua [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    end
+    function TestMap:testAddMapScalar()
+        local r = {{3,4,7},{3.5,5.5,1.3}}
+        local result = Map.add(13, r)
+        local expected = {{3,4,7},{3.5+13,5.5+13,1.3+13}}
+
+        lu.assertItemsEquals(result, expected)
+    end
+    --[[
+    function TestMap:testAddMaps()
```
[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/test/Map-test.lua:35]: Lua unit tests are run as part of the build now! I only wrote a few tests as a "test", but yes, def future todo to unit test all of these very core Map functions

File: client/src/app/expression-language/lua/Map.lua [hunk 3]
```diff
@@ -0,0 +1,387 @@
+local Map = {
```
[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/Map.lua:1]: Good point, will add, thanks!

File: client/src/app/expression-language/lua/Map.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/Map.lua:347]: Doesn't seem used by Mike, it's also not in shared expressions, but I added a normalize that calls normalise anyway for now :)

[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/Map.lua:347]: On second thoughts, Scott did say lets do things with the "queens english", probably less messy to have 1 name for the function if it ever needs transpiling, so I'll just leave it as one normalise() function

File: client/src/app/expression-language/lua/test/MikesLib-test.lua [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        local line, comment = Utils.findMapTableDifferenceLine(result, "result", expected, "expected", pmcsToIgnore)
+
+        if line >= 0 then
+            print("Unexpected result: "..comment.." at line: "..line)
+        else
+            print("Result output was correct, NOTE: IGNORED "..ignoreCount.." values")
+        end
+
+        lu.assertEquals(line, -1)
```
[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/test/MikesLib-test.lua:39]: Was this from running ./test.sh ?
I get all of them passing, that's interesting! I'll have to try run it on the mac too, still haven't given that back. Here's my Windows output:
image

[the remark this answers is not in the document]
Comment (pnemere) [reply client/src/app/expression-language/lua/test/MikesLib-test.lua:39]: This is passing on the build machine too!

**Pull request merged**